File: cache-simulator.py
#!/usr/bin/python

# -----------------------------------------------------------------------------
#  Cache simulation
#  LRU replacement
#  calculate the proportion of BIAS objects in total
# -----------------------------------------------------------------------------

import os
import string
import sys
import linecache

# -----------------------------------------------------------------------------
class g:
   # Global constants
   MEMORY_TIME   = 50 # Number of clocks for read/write
   CACHE_SIZE    = 0  # Number of blocks in cache
   N_WAY         = 2  # For set associative, size of each set

   # Global variables
   cycles        = 0  # Clock cycle count
   inFile        = "" # Filename of input file (cmd line)
   intersizeFile = "" # Filename of inter size file (cmd line)
   addressFile   = "" # Filename of addresses file (cmd line)
   mappingStrat  = 0  # Numeric code for mapping strategy* (cmd line)
   writeStrat    = 0  # Numeric code for write strategy** (cmd line)
   cache         = 0  # Instance of cache class

   hit           = 0  # the total hit numbers
   miss          = 0  # the total miss numbers
   biaslist      = [] # the list used to maintain the bias objects in cache
   biasdict      = {} # the dict contains the bias objects

   # * Mapping Strategy:
   #     1 -> Direct mapped
   #     2 -> Fully associative
   #     3 -> Set associative

   # ** Write Strategy:
   #     1 -> Write through
   #     2 -> Write back

# -----------------------------------------------------------------------------
class Cache:
   """
   Representation of the cache. The cache is represented as a
   dictionary (hash) with the keys being the block numbers and
   the address stored as the value. In addition, there are two
   complementary hashes whose keys are identical to the cache
   blocks. One holds dirty flags for each block, used for the
   write back writing strategy; the other contains LRU values
   for each block, used with the fully associative and set
   associative mapping strategies.
   """
   def __init__(self):
      self.cacheBlocks = {}
      self.dirtyFlags  = {}
      self.lruValue    = {}
      self.clearCache()

   # Zero out cache blocks/flags/values
   def clearCache(self):
      for n in range(g.CACHE_SIZE):
         self.cacheBlocks[n] = 0
         self.dirtyFlags[n]  = 0
         self.lruValue[n]    = 0

   # Check to see if a block is in the cache
   def checkCacheHit(self, address):
      block = self.mapToBlock(address)
      if address == self.cacheBlocks[block]: # Cache hit
         self.incrementLRU(block) # Increase LRU value for this block
         return block
      else:
         return -1

   # Increment specific block LRU value and decrement all
   # others. This represents a simple LRU counting strategy
   def incrementLRU(self, block):
      for n in range(g.CACHE_SIZE): # First, decrement all LRU values
         if self.lruValue[n] > 0 and n != block:
            self.lruValue[n] -= 1

      if self.lruValue[block] < 10: # Then, increment active block (cap is 10)
         self.lruValue[block] += 1

   # Return LRU block from specified slice
   def getLRUBlock(self, dictSlice):
      self.lruSlice = {}

      # Build new dictionary based on slice
      for n in range(dictSlice[0], dictSlice[1]):
         self.lruSlice[n] = self.lruValue[n]

      # Sort LRU values, return first lowest value
      lruValues = self.lruSlice.items()
      lruValues.sort()

      # Sorted list is a list of tuples,
      # so slice out the first of both
      return lruValues[0][0]

   # Store the "data" into the cache
   def storeCacheBlock(self, address, biasdict):
      block = self.mapToBlock(address)

      # Check to see if block is dirty before
      # replacing it. If dirty, write it out to
      # memory before replacement
      if self.isBlockDirty(address) == 1:
         g.cycles += g.MEMORY_TIME
         self.dirtyFlags[block] = 0

      # Return value default is 1
      # If less than 1, means it is the BIAS object's access
      returnvalue = 1
      # Check to see if the block be replaced is biasdict
      if self.cacheBlocks[block] in biasdict:
         if biasdict[self.cacheBlocks[block]] == 0:
            returnvalue = -1
            del biasdict[self.cacheBlocks[block]]

      self.cacheBlocks[block] = address

      return returnvalue

   # Return status of a block's dirty flag
   def isBlockDirty(self, address):
      block = self.mapToBlock(address)

      return self.dirtyFlags[block]

   # Set the dirty flag for a particular address
   def setDirtyFlag(self, address):
      block = self.mapToBlock(address)
      self.dirtyFlags[block] = 1

   # Map memory address to cache block
   def mapToBlock(self, address):
      if g.mappingStrat == 1:   # Direct mapped cache
         return address % g.CACHE_SIZE
      elif g.mappingStrat == 2: # Fully associative cache
         return self.getLRUBlock((0, g.CACHE_SIZE - 1))
      elif g.mappingStrat == 3: # Set associative cache (assume 2-way)
         nWay = g.CACHE_SIZE / g.N_WAY
         set  = address % nWay
         return self.getLRUBlock((set*g.N_WAY, (set*g.N_WAY)+g.N_WAY))

# -----------------------------------------------------------------------------
def readReference(inFile, lineNum):
   """
   Get specified line from input file. The incoming line is
   split into subfields, only two of which are returned. If
   the line does not exist in the file, a zero tuple is
   returned to indicate failure.
   """
   line = linecache.getline(inFile, lineNum)
   if line != '':
      try:
         (rw, addressHex, size) = string.splitfields(line, " ")
         return (rw, int(addressHex, 0))
      except:
         sys.exit("Trace file format: RW-type(0/1) Address Size")
   else:
      return (0, 0)

# -----------------------------------------------------------------------------
def readBias(intersizeFile, addressFile):
   """
   Get the bias objects addresses, only first 0-4 accesses.
   """
   interlist = []
   tmplist = []
   fp = open(intersizeFile, "r")
   for line in fp:
      if line.find("[") == 0:
         tmplist = []
         count = line.count(",")
         if count > 0:
            word = string.split(line, " ")
            tempcoma = string.split(word[0], "[")
            tempaddress = string.split(tempcoma[1], ",")[0]
            tmplist.append(int(tempaddress))
            for n in range(1, count):
               tmplist.append(int(string.split(word[n], ",")[0]))
            temp = string.split(word[count], ']')
            tmplist.append(int(temp[0]))
            interlist.append(tmplist)
         else:
            word = string.split(line, ']')
            tmplist.append(int((string.split(word[0], '['))[1]))
            interlist.append(tmplist)
   fp.close()

   fp = open(addressFile, "r")
   addrlist = []
   for line in fp:
      word = string.split(line, "\n")
      addrlist.append(word[0])
   fp.close()
      
   biasdict = {}
   for i in range(0, 4):
      for j in range(0, len(interlist)):
         try:
            if interlist[j][i] >= 8388608:
               addr = int(addrlist[j], 0)
               biasdict[addr] = i
         except:
            pass

   del interlist[:]
   del addrlist[:]
   del tmplist[:]

   return biasdict

# -----------------------------------------------------------------------------
def printSummary():
   """
   Print results of program execution. This is called at the
   end of the program run to provide a summary of what
   settings were used and the resulting time. In addition,
   the command line number codes are converted to text for
   readability.
   """
   mapping = ""
   write   = ""

   if g.mappingStrat == 1:
      mapping = "Direct mapped"
   elif g.mappingStrat == 2:
      mapping = "Fully associative"
   elif g.mappingStrat == 3:
      mapping = "Set associative"

   if g.writeStrat == 1:
      write = "Write through"
   elif g.writeStrat == 2:
      write = "Write back"

   print "o--------------------------o"
   print "|           Input file:", g.inFile
   print "|       Intersize file:", g.intersizeFile
   print "|         Address file:", g.addressFile
   print "|           Cache size:", repr(g.CACHE_SIZE)
   print "|     Mapping strategy:", mapping
   print "|       Write strategy:", write
   print "|         Clock cycles:", repr(g.cycles)
   print "|           Total hits:", g.hit
   print "|         Total misses:", g.miss
   print "| Most BIAS cache line:", max(g.biaslist)
   print "o--------------------------o"

# =============================================================================
# Everything starts from here
def main():
   """
   Main program; contains primary clock loop and core logic.
   """
   if not len(sys.argv) == 7: # Check for command line arguments
      sys.exit("Usage: %s [ cache size ] [ mapping ] [ write ] [ tracefilename ] [ intersizefilename ] [ addressfilename ] \n" % \
         os.path.basename(sys.argv[0]))

   # Read input file from command line
   g.CACHE_SIZE    = int(sys.argv[1]) # Cache size
   g.mappingStrat  = int(sys.argv[2]) # Cache mapping (1, 2, 3)
   g.writeStrat    = int(sys.argv[3]) # Cache write strategy (1, 2)
   g.inFile        = sys.argv[4]      # Input text file
   g.intersizeFile = sys.argv[5]      # Inter Size file
   g.addressFile   = sys.argv[6]      # Address file

   # Initialize variables
   g.cache  = Cache() # Set instance of cache class
   rw       = ""      # Reference read/write flag
   address  = 0       # Reference address
   line     = 1       # Current line number in input file
   cnt      = 0       # Count of cache lines having BIAS objects in cache
   temp     = 0       # Use this number to compare the cnt number, reduce the memory usage to record biaslist

   # read the bias objects along with occurance index
   g.biasdict = readBias(g.intersizeFile, g.addressFile)

   # Core logic
   while 1:
      # Get references --------------------------------------------------------
      (rw, address) = readReference(g.inFile, line)
      line += 1
      temp = cnt
      if address == 0:
         break # EOF, so shut down

      # For speed, make these function calls once so
      # we can just test the return values
      cacheHit = g.cache.checkCacheHit(address)

      if address in g.biasdict:
         if g.biasdict[address] > 0:
            g.biasdict[address] -= 1
         else:
            cnt += 1
         #   del g.biasdict[address]

      # Read reference --------------------------------------------------------
      if rw == '0':
         if cacheHit > 0:
            # Data is in cache, so move on
            g.hit += 1
            g.cycles += 1
         else:
            # Data is not in cache, so pull it from
            # lower memory and then store it in the
            # cache
            g.miss += 1
            g.cycles += g.MEMORY_TIME
            if g.cache.storeCacheBlock(address, g.biasdict) < 0:
               cnt -= 1

      # Write reference -------------------------------------------------------
      elif rw == '1':
         if g.writeStrat == 1: # <<<----------------------- Write through
            if cacheHit > 0:
               # Write data to block in lower memory
               g.hit += 1
               g.cycles += g.MEMORY_TIME + 1
            else:
               # Data is not in cache, so add it to the
               # cache and then update memory
               g.miss += 1
               if g.cache.storeCacheBlock(address, g.biasdict) < 0:
                  cnt -= 1
               g.cycles += g.MEMORY_TIME + 1
         elif g.writeStrat == 2: # <<<--------------------- Write back
            if cacheHit > 0:
               # Flag block as dirty; this data will
               # be written back out to memory on replacement
               g.hit += 1
               g.cycles += 1
               g.cache.setDirtyFlag(address)
            else:
               # Data is not in cache, so write it to the
               # cache and then flag the block as dirty
               g.miss += 1
               if g.cache.storeCacheBlock(address, g.biasdict) < 0:
                  cnt -= 1
               g.cycles += 1
               g.cache.setDirtyFlag(address)

   # Update the BIAS cache line list
   if temp != cnt:
      g.biaslist.append(cnt)
   # Display results of program run
   printSummary()

if __name__ == "__main__": 
   main()

File: C++ ver/cache-simulator.cc
/* cache-simulator.cc
 *
 */

#include <unordered_map>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#define memory_time  50   // cycles used to commute with memory
#define n_way  2          // if it is associative, size of each set

using namespace std;

int cycles=0;
int hit=0;
int miss=0;
int cachesize=8388608;            // cache blocks
int mappingStrat=3;

typedef boost::unordered_map<long, int> map;

class Cache {
    private:
        map cacheBlocks;
        map diryFlags;
        map lruValue;
        map dictSlice;
    public:
        void _init_(){
            clearcache();
        } // end initialization

        void clearcache(){
            for (int i=0; i<cachesize; i++){
                cacheBlocks[i] = 0;
                dirtyFlags[i] = 0;
                lruValue[i] = 0;
            }
        } // end clearcache function 

        int checkCacheHit(long address){
            int block = mapToBlock(address);
            if (address == cacheBlocks[block]){
                incrementLRU(block);
                return block;
            } else
                return -1;
        } // end checkCacheHit functon

        void incrementLRU(int block){
            for (int i=0; i<cachesize; i++)
                if (lruValue[i] > 0 && i!=block)
                    lruValue[i]--;
            
            if (lruValue[block] < 10)
                lruValue[block]++;
        } // end incrementLRU function

        int getLRUBlock(int start, int end){
            map lruSlice;
            int lruValues = 10000;
            for (int i=start; i<end; i++){
                lruSlice[i] = lruValue[i];
                if (lruValues > lruSlice[i])
                    lruValues = lruSlice[i];
            }
            return lruValues;
        } // end getLRUBlock

        int storeCacheBlock(long address, map biasdict){
            int blcok = mapToBlock(address);
            if (isBlockDirty(address) == 1){
                cycles += memory_time;
                dirtyFlags[block] = 0;
            }

            int returnvalue = 1;
            if (biasdict.find(cacheBlocks[block]) != biasdict.end){
                if (biasdict[cacheBlocks[block]] == 0){
                    returnvalue = -1;
                    biasdict.erase(cacheBlocks[block]);
                }
            } // end if

            cacheBlocks[block] = address;
            return returnvalue;
        } // end store cache block function

        int isBlockDirty(long address){
            int block = mapToBlock(address);
            return dirtyFlags[block];
        }

        void setDirtyFlag(long address){
            int block = mapToBlock(address);
            dirtyFlags[block] = 1;
        }

        int mapToBlock(long address){
            if (mappingStrat == 1)
                return address%cachesize;
            else if (mappingStrat == 2)
                return getLRUBlock(0, (cachesize-1));
            else if (mappingStrat == 3){
                int nway = cachesize/n_way;
                int set = address%nway;
                return getLRUBlock(set*n_way, (set*n_way)+n_way);
            }
        } // end map to block function
};

@inline
int stoi(string str){
    stringstream ss(str);
    int res;

    if (!ss>>res){
        cout << "error happens when translet string: " << str << endl;
        return -1;
    }

    return res;
}

int readBias(string intersizefile, string addressfile, unordered_map<long,int> &biasdict){
    ifstream infile(addressfile);
    if (!infile.is_open()){
        cout << "No address file exist!\n";
        return -1;
    }

    long address;
    vector<long> addr;
    while(infile>>address)
        addr.push_back(address);

    ifstream fa(intersizefile);

    if (!fa.is_open()){
        cout << "No Intersize file exist!\n";
        return -1;
    }

    string line;
    int i=0;
    while(getline(fa, line)){
        int start=1;
        int end=1;
        int size = 0;
        int cnt=0;
        while(end=line.find(',',end)){
            string sub = line.sutbstr(start, end-start+1);
            size = stoi(sub);
            start=end;
            if (size>8388608){
                biasdict[addr[i]]=cnt;
                break;
            }
            cnt++;
        }
        i++;
    } // end while loop
    return 0;
}

@inline
void helper(){
    cout << "Cache simulator calculating the percentage of BIAS objects in the cache\n";
    cout << "Usage: cachesim trace-file-path address-file-path inter-size-file-path\n";
    cout << "Cache parameters are defined in file already. \n";
    cout << "Commented the corresponding lines out, if parameters are assigned in CMDLine\n";
}

// everything starts from here
int main(int argc, char ** argv){
    if (argc<5){
        helper();
        return 0;
    }

    string tracefile     = *(argv+1);
    string addressfile   = *(argv+2);
    string intersizefile = *(argv+3);
    long cnt              = 0;  // count the cache lines having bias objects
    vector<long> count;         // vector used to store the cnt

    unordered_map<long, int> biasdict;
    if(!readBias(intersizefile, addressfile, biasdict)){
        cout << "Failed to read the bias\n";
        helper();
        return 0;
    }

    Cache *cache = new Cache();
    cache->_init_();

    ifstream infile(tracefile);
    if (!infile.is_open()){
        cout << "Failed to open trace file\n";
        helper();
        return 0;
    }

    long address;
    int mode;
    int size;
    while(infile>>address>>mode>>size){
        int cachehit;
        cachehit = cache->checkCacheHit(address);

        if (biasdict.find(address) != biasdict.end()){
            if (biasdict[address]>0)
                biasdict[address]--;
            else{
                biasdict[address]--;
                cnt++;
                count.push_back(cnt);
            }
        } // end if bias disct has this key

        if (mode==0){
            if (cachehit>0){
                hit++;
                cycles++;
            } // end hit else
            else{
                miss++;
                cycles+=memory_time;
                if (cache->storeCacheBlock(address, biasdict)<0){
                    cnt--;
                    count.push_back(cnt);
                } // end if
            }  //end else miss
        } // end read if

        else if (mode==1){
            if (writeStrat==1){
                if (cachehit>0){
                    hit++;
                    cycles+=memory_time;
                }
                else{
                    miss++;
                    cycles+=memory_time;
                    if(cache->storeCacheBlock(address,biasdict)<0){
                        cnt--;
                        count.push_back(cnt);
                    }
                }
            } // end write through strategy
            else if(writeStrat==2){
                if (cachehit>0){
                    hit++;
                    cycles+=1;
                    cache->setDirtyFlag(address);
                } else {
                    miss++;
                    cycles+=1;
                    if (cache->storeCacheBlock(address,biasdict)<0){
                        cnt--;
                        count.push_back(cnt);
                    }
                    cache->setDirtyFlag(address);
                }
            } // end write back strategy
        } // end write if
    } // end file reading in

    infile.close();

    ofstream outfile("bias-occupation-cache");
    for(int ii=0;ii<count.size();ii++)
        outfile<<count[ii]<<endl;
    outfile.close();
    return 0;
}